### PR TITLE
Removed Support for Python2

### PR DESCRIPTION
**Pull request by ashah-splunk**
- removed code specific for python version2
- removed six.py dependency which bridged support for python versions 2 and 3
- refactored code as per pylint suggestions
- used python3 features to improve both efficiency and performance
- removed future imports required to support py2 and py3 versions
- updated test cases to use py3 code
- updated CI test configuration to execute test cases in python 3.7 and 3.9 versions
- updated code to remove code deprecation warnings

### CODE REVIEW

File: .github/workflows/test.yml
Comment (shilpabijam) [inline .github/workflows/test.yml:14]: If I recall we discussed adding versions of python that currently ship with Enterprise?

**Comment by shilpabijam** [comment]
There are some test failures in latest and 8.1 runs, could we address them?

File: CHANGELOG.md
Comment (shilpabijam) [inline CHANGELOG.md:14]: Seeing some deprecation warnings in the test runs, are these expected?

File: splunklib/__init__.py
Comment (shilpabijam) [inline splunklib/__init__.py:33]: These seem like utility methods that can reside in the utils package?

Comment (shilpabijam) [inline splunklib/__init__.py:33]: I would think a better place to put these utility methods would be the utils package .

File: splunklib/searchcommands/internals.py
Comment (jervin-splunk) [inline splunklib/searchcommands/internals.py:54]: Since we are removing support for Python 2 entirely in this MR, this method needs more simplification to remove the checks for sys.version_info - these are no longer necessary.

File: tests/test_job.py
Comment (jervin-splunk) [inline tests/test_job.py:30]: There's code on line 33 below which refers to Python 2.6 - recommend determining whether the comment is misleading and can now be removed, or whether more work is needed to catch ParseErrors as well.

**Review by jervin-splunk — CHANGES_REQUESTED** [review]
In addition to the places noted in the other comments, there are some lingering references to Python 2.7 and 2.6 which should likely be cleaned up as well in these files:

```
splunk-sdk-python/tests/README.md
splunk-sdk-python/scripts/test_specific.sh
splunk-sdk-python/README.md
```

File: Makefile
Comment (ashah-splunk) [reply Makefile:37]: this was not intentional, it was missed from our side

File: splunklib/searchcommands/external_search_command.py
Comment (maszyk99) [inline splunklib/searchcommands/external_search_command.py:37]:
```suggestion
        if not isinstance(path, (bytes, str)):
```

File: .github/workflows/test.yml
Comment (ashah-splunk) [reply .github/workflows/test.yml]: as per the discussion, python 3.7 and Python 3.9 will be part of the matrix.

File: CHANGELOG.md
Comment (ashah-splunk) [reply CHANGELOG.md:14]: yes those warnings are expected as they are regarding couple of features that we have replaced but have to decide on when to remove them permanently from the code.

File: tests/searchcommands/chunked_data_stream.py
Comment (ashah-splunk) [reply tests/searchcommands/chunked_data_stream.py]: added the method here

File: tests/searchcommands/chunked_data_stream.py
Comment (ashah-splunk) [reply tests/searchcommands/chunked_data_stream.py]: yes, we missed this. Used `ensure_binary()` method here

File: tests/test_collection.py
Comment (ashah-splunk) [reply tests/test_collection.py]: rectified the missing colon

File: tests/test_collection.py
Comment (ashah-splunk) [reply tests/test_collection.py]: rectified the missing colons

File: tests/test_collection.py
Comment (ashah-splunk) [reply tests/test_collection.py]: rectified the missing colons.

File: tests/searchcommands/chunked_data_stream.py
Comment (ashah-splunk) [reply tests/searchcommands/chunked_data_stream.py]: yes missed the individual encodes in the code, added `ensure_binary()` method here

File: splunklib/__init__.py
Comment (ashah-splunk) [reply splunklib/__init__.py]: methods have been moved to utils

File: CHANGELOG.md
Comment (ashah-splunk) [reply CHANGELOG.md]: yes, only 3.7 and 3.9

File: splunklib/client.py
Comment (ashah-splunk) [reply splunklib/client.py]: modified the code as per the feedback

File: tests/test_job.py
Comment (ashah-splunk) [reply tests/test_job.py:30]: we had a look and it seems safe to just remove the comments.

File: splunklib/searchcommands/internals.py
Comment (ashah-splunk) [reply splunklib/searchcommands/internals.py]: yes, simplified the method and removed the unnecessary code

File: tests/searchcommands/test_internals_v2.py
Comment (ashah-splunk) [reply tests/searchcommands/test_internals_v2.py]: these change may have been added for some testing purpose, not recollecting the reason for adding these.
Have removed this change after analysing if they have any impact on test execution.

File: tests/searchcommands/test_decorators.py
Comment (ashah-splunk) [reply tests/searchcommands/test_decorators.py]: this change may have been added for some testing purpose, not recollecting the reason for adding these.
Have removed this change after analysing if they have any impact on test execution.

File: tests/searchcommands/chunked_data_stream.py
Comment (ashah-splunk) [reply tests/searchcommands/chunked_data_stream.py]: rectified the mistake.

File: splunklib/searchcommands/search_command.py
Comment (ashah-splunk) [reply splunklib/searchcommands/search_command.py:972]: To add access for finished flag within metadata, update method was added 
code commit Ref - https://github.com/splunk/splunk-sdk-python/pull/456/commits/bf40010ed603c1f35a72ea9b437f0c672024c8e3

File: splunklib/searchcommands/external_search_command.py
Comment (ashah-splunk) [reply splunklib/searchcommands/external_search_command.py:37]: will check for formatting in each file and commit the changes together

File: splunklib/searchcommands/internals.py
Comment (jervin-splunk) [inline splunklib/searchcommands/internals.py:690]: If we are not cleaning up the code which purports to support PyPy 2.5 - which we *definitely* do not support at this stage given that that version was released in 2015 - then we should reinstate this comment, otherwise the code does not make sense.

Ideally, this entire block would be refactored to  modern standards; importing the private implementation of the module by using the underscore name should really no longer be necessary.

**Review by ichaer — CHANGES_REQUESTED** [review]

File: splunklib/binding.py
Comment (maszyk99) [reply splunklib/binding.py:123]: I fixed it for files that were edited in this PR. I will create a task to remove it from all other places.

File: splunklib/client.py
Comment (maszyk99) [reply splunklib/client.py:2348]: @ashah-splunk We will need your assitance here

File: splunklib/searchcommands/internals.py
Comment (maszyk99) [reply splunklib/searchcommands/internals.py:793]: @ashah-splunk

File: splunklib/client.py
Comment (maszyk99) [reply splunklib/client.py:3476]: Ideally we should split it into another pull request and review it just after this one gets merged, but maybe this change is needed as fast as possible. What do you think @ashah-splunk ?

Comment (ashah-splunk) [reply splunklib/client.py:3476]: Hi @ichaer,
this change was made in reference to the issue https://github.com/splunk/splunk-sdk-python/issues/527.
As per the discussions with the team then, we decided to add this as a part of this PR. 
We have tested the code as well as its been reviewed by the team then before pushed to the PR.

File: splunklib/searchcommands/search_command.py
Comment (ashah-splunk) [reply splunklib/searchcommands/search_command.py:972]: Hi @ichaer ,
This change was made in reference to the slack thread - https://splunk.slack.com/archives/CA9R2GU4U/p1669930999509219
It was reported by some customer so it was considered to be added in this PR. At that time plan was in-motion to release the Python 2.0.0-beta version, but then the Org restructure happened and the release was postponed.

File: splunklib/searchcommands/internals.py
Comment (ashah-splunk) [reply splunklib/searchcommands/internals.py:793]: Hi @ichaer 
To my knowledge this change was made in reference to the Jira ticket - https://splunk.atlassian.net/browse/DVPL-11850.
As per the comments on the ticket and response by Navya - https://splunk.atlassian.net/browse/DVPL-11850?focusedCommentId=10596736
This was at that time planned as part of the refactor hence added to this PR.

File: splunklib/client.py
Comment (ichaer) [reply splunklib/client.py:3476]: OK, I won't block you on this, but I would urge you to review your change process for future PRs. PRs are cheap to create and merge, there's no reason to group unrelated changes into one PR, and splitting them allows for more focused reviews and testing, as well as easier reverts.

File: splunklib/searchcommands/internals.py
Comment (ichaer) [reply splunklib/searchcommands/internals.py:793]: OK, if we want to keep the change let's remove the commented-out code. History is already preserved by git, let's leave it out of the code.

File: splunklib/client.py
Comment (ashah-splunk) [reply splunklib/client.py]: sure @ichaer , we will make sure to have separate PRs going further. Thanks!

File: splunklib/searchcommands/internals.py
Comment (ashah-splunk) [reply splunklib/searchcommands/internals.py]: Ok, we have removed the commented out code.

File: splunklib/client.py
Comment (ashah-splunk) [reply splunklib/client.py]: @ichaer we have added the `else` block back.

File: splunklib/searchcommands/internals.py
Comment (maszyk99) [reply splunklib/searchcommands/internals.py]: I tried to refactor entire block to modern standards by simply adding function with `json.dumps`, but there are some corner cases that this method doesn't cover and it causes one test to fail.

For the sake of the release, I just added the comment back for now.

Comment (jervin-splunk) [reply splunklib/searchcommands/internals.py]: Sounds good.

File: splunklib/searchcommands/external_search_command.py
Comment (ichaer) [reply splunklib/searchcommands/external_search_command.py]: You haven't acted on this, the log line is still wrong.

File: splunklib/searchcommands/search_command.py
Comment (ichaer) [reply splunklib/searchcommands/search_command.py]: This hasn't been addressed either.

**Comment by ichaer** [comment]
I'm approving anticipating that the two pending comments will be resolved. They are not about particularly concerning issues, but they are trivial to implement and shouldn't be postponed.

**Review by ichaer — APPROVED** [review]

File: splunklib/searchcommands/search_command.py
Comment (ashah-splunk) [reply splunklib/searchcommands/search_command.py]: @ichaer this change has been applied, I think we missed to update here.
commit ref - https://github.com/splunk/splunk-sdk-python/pull/456/commits/9053482a3ef97e9d04684fec9d4117f1e40f6394

File: splunklib/searchcommands/external_search_command.py
Comment (ashah-splunk) [reply splunklib/searchcommands/external_search_command.py]: @ichaer we have resolved this as well. please find the commit here - https://github.com/splunk/splunk-sdk-python/pull/456/commits/1d1cc207f131682e02773f54df80960def228245.
sorry we missed to update here.

**Pull request merged**
